src/ConnectedMode/Migration/ConnectedModeMigration.cs

```diff
@@ -42,19 +44,53 @@ private sealed class ChangedFiles : List<FilePathAndContent<string>> { }
         private readonly IFileProvider fileProvider;
         private readonly IFileCleaner fileCleaner;
         private readonly IVsAwareFileSystem fileSystem;
+        private readonly ISonarQubeService sonarQubeService;
         private readonly ILogger logger;
 
+        // The user can have both the legacy and new connected mode files. In this case the behaviour is slightly different.
+        private bool isConnectedToNewConnectedMode;
+
         [ImportingConstructor]
-        public ConnectedModeMigration(IMigrationSettingsProvider settingsProvider, IFileProvider fileProvider, IFileCleaner fileCleaner, IVsAwareFileSystem fileSystem, ILogger logger)
+        public ConnectedModeMigration(IMigrationSettingsProvider settingsProvider,
+            IFileProvider fileProvider,
+            IFileCleaner fileCleaner,
+            IVsAwareFileSystem fileSystem,
+            ISonarQubeService sonarQubeService,
+            ILogger logger)
         {
             this.settingsProvider = settingsProvider;
             this.fileProvider = fileProvider;
             this.fileCleaner = fileCleaner;
             this.fileSystem = fileSystem;
+            this.sonarQubeService = sonarQubeService;
             this.logger = logger;
         }
 
         public async Task MigrateAsync(BoundSonarQubeProject oldBinding, IProgress<MigrationProgress> progress, CancellationToken token)
+        {
+            isConnectedToNewConnectedMode = sonarQubeService.IsConnected;
+
+            if (!isConnectedToNewConnectedMode)
+            {
+               await sonarQubeService.ConnectAsync(oldBinding.CreateConnectionInformation(), token);
+            }
+
+            try
+            {
+                await MigrateImplAsync(oldBinding, progress, token);
+            }
+            catch (Exception ex) when (!ErrorHandler.IsCriticalException(ex))
+            {
+                if (!isConnectedToNewConnectedMode)
```

Comment:
It's probably worth adding a comment here e.g. `If we establish the server connection during migration, disconnect it if the migration failed.`